pkg/resources/secret_with_oauth_client_credentials.go

```diff
@@ -0,0 +1,212 @@
+
+	stringScopes := expandStringList(d.Get("oauth_scopes").(*schema.Set).List())
+	oauthScopes := make([]sdk.ApiIntegrationScope, len(stringScopes))
+	for i, scope := range stringScopes {
+		oauthScopes[i] = sdk.ApiIntegrationScope{Scope: scope}
+	}
+	request.WithOauthScopes(sdk.OauthScopesListRequest{OauthScopesList: oauthScopes})
+
+	if v, ok := d.GetOk("comment"); ok {
+		request.WithComment(v.(string))
+	}
+
+	err = client.Secrets.CreateWithOAuthClientCredentialsFlow(ctx, request)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(helpers.EncodeResourceIdentifier(id))
+
+	return ReadContextSecretWithClientCredentials(ctx, d, meta)
+}
+
+func ReadContextSecretWithClientCredentials(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	client := meta.(*provider.Context).Client
+	id, err := sdk.ParseSchemaObjectIdentifier(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	secret, err := client.Secrets.ShowByID(ctx, id)
+	if err != nil {
+		if errors.Is(err, sdk.ErrObjectNotFound) {
+			d.SetId("")
+			return diag.Diagnostics{
+				diag.Diagnostic{
+					Severity: diag.Warning,
+					Summary:  "Failed to retrieve secret with client credentials. Target object not found. Marking the resource as removed.",
+					Detail:   fmt.Sprintf("Secret with client credentials name: %s, Err: %s", id.FullyQualifiedName(), err),
+				},
+			}
+		}
+		return diag.Diagnostics{
+			diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  "Failed to retrieve secret with client credentials.",
+				Detail:   fmt.Sprintf("Id: %s\nError: %s", d.Id(), err),
+			},
+		}
+	}
+	secretDescription, err := client.Secrets.Describe(ctx, id)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err := handleSecretRead(d, id, secret, secretDescription); err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err = d.Set("api_authentication", secretDescription.IntegrationName); err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err = d.Set("oauth_scopes", secretDescription.OauthScopes); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func UpdateContextSecretWithClientCredentials(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	client := meta.(*provider.Context).Client
+	id, err := sdk.ParseSchemaObjectIdentifier(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	commonSet, commonUnset := handleSecretUpdate(d)
+	set := &sdk.SecretSetRequest{
+		Comment: commonSet.comment,
+	}
+	unset := &sdk.SecretUnsetRequest{
+		Comment: commonUnset.comment,
+	}
+
+	if d.HasChange("oauth_scopes") {
+		stringScopes := expandStringList(d.Get("oauth_scopes").(*schema.Set).List())
+		oauthScopes := make([]sdk.ApiIntegrationScope, len(stringScopes))
+		for i, scope := range stringScopes {
+			oauthScopes[i] = sdk.ApiIntegrationScope{Scope: scope}
+		}
+		req := sdk.NewSetForOAuthClientCredentialsRequest().WithOauthScopes(*sdk.NewOauthScopesListRequest(oauthScopes))
```

Comment:
question: it works on the Snowflake side that it replaces the previous scopes or does it leave the previous ones too?

Reply:
In Snowflake it replaces previous scope to empty list e.g. "[foo bar]" to "[]"